### DISCUSSION TITLE
Why "C[O]" does not canonicalize to "CO"?

### DISCUSSION THREAD

**Discussion by rczerminski-valo** [rdkit/rdkit#5605] [Q&A]
I would assume that ```C[O]``` SMILES does represent exactly the same molecular graph as ```[CH3][OH]```, but it does not seem to be the case. Why? And, if at all possible, how to generate the same canonical SMILES string from these different strings?

```
>>> sm = ["[CH3][OH]", "CO", "[C][O]", "C[O]", "[C]O"]
>>> mols = [MolFromSmiles(x) for x in sm]
>>> [MolToSmiles(x) for x in mols]
['CO', 'CO', '[C][O]', 'C[O]', '[C]O']
>>>
```

Or possibly C[O] is interpreted as a free-radical in RDKit i.e. more explicitly as [CH3][#8&H0]. In such a case, my question is what is the simplest way to convert it (and in general free-radicals) to "regular" molecules?

**Comment by mwojcikowski** [comment]
These are two different molecules. More specifically `C[O]` contains a radical oxygen. SMILES notation in the square brackets needs Hs to be explicitly stated. If there isn't any, it means no hydrogen `H0`. So `C[O]` is `[CH3][O]`, and `CO` is `[CH3][OH]`.

**Reply by rczerminski-valo** [reply]
Yes, I realized this shortly after posting the original question :) As a follow up: Is there some function to convert radical -> non-radical? I just cooked up one below, but maybe there is something better existing already.

```
def radical2mol(mol, debug=False):
    """
    Check for free radicals and if found any add Hs
    """
    k = 0
    for a in mol.GetAtoms():
        nr = a.GetNumRadicalElectrons()
        ne = a.GetNumExplicitHs()
        ni = a.GetNumImplicitHs()
        nh = nr + ne + ni # total number of hydrogens
        if debug: print(f"DEBUG: radical2mol: k={k}, nr={nr}, ne={ne}, ni={ni}, nh={nh}")
        if nr > 0:
            a.SetNumExplicitHs(nh)
            a.SetNumRadicalElectrons(0)
        k += 1
    return None
```

**Reply by mwojcikowski** [reply]
Your method is fine. You could also make them implicit by recalculating the caches, but I guess your code would be faster for this specific use case.
